src/query/sql/src/planner/optimizer/decorrelate/subquery_rewriter.rs

```diff
@@ -452,6 +460,169 @@ impl SubqueryRewriter {
+                Self {
+                    replacer: BindingReplacer {
+                        new_column_bindings,
+                    },
+                    failed: false,
+                    predicates: Vec::new(),
+                }
+            }
+
+            fn visit(&mut self, s_expr: &SExpr) -> Result<()> {
+                for visit in s_expr.children() {
+                    self.visit(visit)?;
+                }
+                if self.failed {
+                    return Ok(());
+                }
+                match s_expr.plan() {
+                    RelOperator::Scan(_) | RelOperator::Sort(_) | RelOperator::EvalScalar(_) => (),
+                    RelOperator::Filter(filter) => {
+                        let mut filter = filter.clone();
+
+                        for scalar_expr in filter.predicates.iter_mut() {
+                            self.replacer.visit(scalar_expr)?;
+                        }
+                        self.predicates.append(&mut filter.predicates);
+                    }
+                    _ => {
+                        self.failed = true;
+                        return Ok(());
+                    }
+                }
+                Ok(())
+            }
+
+            fn result(self) -> Option<ScalarExpr> {
+                (!self.failed).then(|| {
+                    self.predicates.into_iter().fold(
+                        ScalarExpr::ConstantExpr(ConstantExpr {
+                            span: None,
+                            value: Scalar::Boolean(true),
+                        }),
+                        |acc, expr| {
+                            ScalarExpr::FunctionCall(FunctionCall {
+                                span: None,
+                                func_name: "and".to_string(),
+                                params: vec![],
+                                arguments: vec![acc, expr],
+                            })
+                        },
+                    )
+                })
+            }
+        }
+
+        #[derive(Debug)]
+        struct BindingReplacer {
+            new_column_bindings: HashMap<IndexType, ColumnBinding>,
+        }
+
+        impl VisitorMut<'_> for BindingReplacer {
+            fn visit_bound_column_ref(&mut self, col: &mut BoundColumnRef) -> Result<()> {
+                if let Some(new_binding) = self.new_column_bindings.get(&col.column.index) {
+                    col.column = new_binding.clone();
+                }
+                Ok(())
+            }
+        }
+
+        let right_expr_binding = &subquery.output_column;
+        let (SubqueryType::Any, Some(box ScalarExpr::BoundColumnRef(left_column))) =
+            (&subquery.typ, &subquery.child_expr)
+        else {
+            return Ok(None);
+        };
+        // `xx in null` != `where xx = null`
+        if left_column.column.data_type.is_nullable() || !is_conjunctive_predicate {
```

Comment:
if data type of left column is nullable, we can still apply this optimization by adding a filter expr `xx is not null`